plugin/trino-hive/src/main/java/io/trino/plugin/hive/util/PartitionMatchSupplier.java

```diff
@@ -0,0 +1,100 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.hive.util;
+
+import io.trino.plugin.hive.HiveColumnHandle;
+import io.trino.plugin.hive.HivePartition;
+import io.trino.spi.connector.DynamicFilter;
+
+import javax.annotation.Nullable;
+
+import java.util.List;
+import java.util.function.BooleanSupplier;
+
+import static io.trino.plugin.hive.HivePartitionManager.partitionMatches;
+import static java.util.Objects.requireNonNull;
+
+public final class PartitionMatchSupplier
+        implements BooleanSupplier
+{
+    private final DynamicFilter dynamicFilter;
+    private final HivePartition hivePartition;
+    private final List<HiveColumnHandle> partitionColumns;
+
+    @Nullable
+    private volatile Boolean finalResult; // value is null until the dynamic filter no longer needs to be evaluated
+
+    private PartitionMatchSupplier(DynamicFilter dynamicFilter, HivePartition hivePartition, List<HiveColumnHandle> partitionColumns)
+    {
+        this.dynamicFilter = requireNonNull(dynamicFilter, "dynamicFilter is null");
+        this.hivePartition = requireNonNull(hivePartition, "hivePartition is null");
+        this.partitionColumns = requireNonNull(partitionColumns, "partitionColumns is null");
+    }
+
+    @Override
+    public boolean getAsBoolean()
+    {
+        // Check for a final memoized result
+        Boolean result = this.finalResult;
+        if (result != null) {
+            return result;
+        }
+        return evaluateCurrentDynamicFilter();
+    }
+
+    private boolean evaluateCurrentDynamicFilter()
+    {
+        // Must be checked before calling dynamicFilter.getCurrentPredicate()
+        boolean filterIsComplete = dynamicFilter.isComplete();
+        boolean partitionMatches = partitionMatches(partitionColumns, dynamicFilter.getCurrentPredicate(), hivePartition);
```

Comment:
reinforce the message about ordering of dynamicFilter calls:

```suggestion
       currentPredicate = dynamicFilter.getCurrentPredicate();
       
        boolean partitionMatches = partitionMatches(partitionColumns, currentPredicate, hivePartition);
```

Reply:
Done